lib/ide/src/controller/text.rs

```diff
@@ -0,0 +1,118 @@
+//! Text Controller module.
+//!
+//! Facade over filesystem API or module text API for text editor. Does discerning between Luna
+//! module file and plain text file. In case of luna module idmap and metadata are hidden for the
+//! user.
+
+use crate::prelude::*;
+
+use crate::controller::*;
+
+use failure::_core::fmt::{Formatter, Error};
+use flo_stream::{Publisher, Subscriber};
+use flo_stream::MessagePublisher;
+use file_manager_client as fmc;
+use json_rpc::error::RpcError;
+use shapely::shared;
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// A notification from TextController.
+#[derive(Clone,Debug)]
+pub enum Notification {
+    /// File contents needs to be set to the following due to synchronization with external state.
+    SetNewContent(String),
+}
+
+
+
+// =======================
+// === Text Controller ===
+// =======================
+
+shared! { ControllerHandle
+
+    /// Data stored by the text controller.
+    pub struct State {
+        /// A path to file the controller is handling.
+        file_path : fmc::Path,
+        /// A module handle in case the TextController is managing Luna module file.
+        module: Option<module::ControllerHandle>,
+        /// Sink where we put events to be consumed by the view.
+        notification_publisher: Publisher<Notification>,
+        /// File manager handle, used to obtain file information in case of plain text file.
+        file_manager: fmc::ClientHandle,
+    }
+
+    impl {
+        /// Create controller managing plain text file.
+        pub fn new(path:fmc::Path, file_manager:fmc::ClientHandle) -> Self {
+            Self {file_manager,
+                file_path              : path,
+                module                 : None,
+                notification_publisher : Publisher::new(10),
+            }
+        }
+
+        /// Get subscriber receiving controller's notifications.
+        pub fn subscribe(&mut self) -> Subscriber<Notification> {
+            self.notification_publisher.subscribe()
+        }
+
+        /// Get clone of file path handled by this controller.
+        pub fn file_path_clone(&self) -> fmc::Path {
+            self.file_path.clone()
+        }
+    }
+}
+
+impl ControllerHandle {
+    /// Create controller managing Luna module file.
+    pub fn new_for_module(module:module::ControllerHandle, file_manager:fmc::ClientHandle) -> Self {
+        let file_path = module.location_as_path();
+        let mut state = State::new(file_path, file_manager);
+        state.module = Some(module);
+        Self { rc:Rc::new(RefCell::new(state)) }
+    }
+
+    /// Read file's content.
+    pub fn read_content(&self) -> impl Future<Output=Result<String,RpcError>> {
```

Comment:
Well, but this function is designed for both kind of files; the reason it does not check for file being module is simply because the reading module content is not implemented.

Reply:
Added match branch causing unimplemented error.